sqlair.go

```diff
@@ -369,7 +500,31 @@ func (tx *TX) Query(ctx context.Context, s *Statement, inputArgs ...any) *Query
 	if ctx == nil {
 		ctx = context.Background()
 	}
+	if tx.isDone() {
+		return &Query{ctx: ctx, err: ErrTXDone}
+	}
 
 	qe, err := s.pe.Query(inputArgs...)
-	return &Query{qs: tx.tx, qe: qe, ctx: ctx, err: err}
+	if err != nil {
+		return &Query{ctx: ctx, err: err}
+	}
+
+	// A transaction is tied to a specific database connection. When a
+	// Statement is prepared on a database it requires a free connection that
+	// does not have a transaction in progress. If all connections are busy
+	// then it is possible to enter a deadlock.
+	// When all connections are busy, prepare on the transaction rather than
+	// the DB.
+	var sqlstmt *sql.Stmt
+	dbStats := tx.db.sqldb.Stats()
+	if dbStats.MaxOpenConnections > 0 && dbStats.MaxOpenConnections == dbStats.InUse {
+		sqlstmt, err = tx.sqltx.PrepareContext(ctx, s.pe.SQL())
```

Comment:
There's clearly a race here. Let's chat about this, please.

Reply:
Fixed by adding an explicit connection underneath the transaction.